codegen/type_converter.go

```diff
@@ -638,25 +671,78 @@ func (c *TypeConverter) genStructConverter(
 			var (
 				stFromPrefix = keyPrefix
 				stFromType   compile.TypeSpec
+				fromTypeName string
 			)
 			if fromField != nil {
 				stFromType = fromField.Type
 				stFromPrefix = keyPrefix + pascalCase(fromField.Name)
+
+				var err error
+				fromTypeName, err = c.getIdentifierName(stFromType)
+				if err != nil {
+					fromTypeName = ""
+				}
 			}
-			err := c.genConverterForStruct(
-				toField.Name,
-				toFieldType,
-				toField.Required,
-				stFromType,
-				fromIdentifier,
-				keyPrefix+pascalCase(toField.Name),
-				stFromPrefix,
-				indent,
-				fieldMap,
-			)
+
+			toTypeName, err := c.getIdentifierName(toFieldType)
 			if err != nil {
 				return err
 			}
+
+			if converterMethodName, ok := c.convStructMap[toFieldType.Name]; ok {
+				// the converter for this struct already exists, so just use it
+				c.append(indent, "out.", keyPrefix+pascalCase(toField.Name), " = ", converterMethodName, "(", fromIdentifier, ")")
+			} else if fromTypeName != "" {
+				// generate a converter for this scope
+				converterMethodName := "convert" + toFieldType.Name + "Helper"
+				c.convStructMap[toFieldType.Name] = converterMethodName
+
+				c.append(indent, "{") // begin scope
+				funcSignature := "func(in *" + fromTypeName + ") (out *" + toTypeName + ")"
+
+				c.append(indent+"\t", "var ", converterMethodName, " ", funcSignature) // function declaration
+				c.append(indent+"\t", converterMethodName, " = ", funcSignature, " {") // begin function definition
+
+				err = c.genConverterForStruct(
+					toField.Name,
+					toFieldType,
+					toField.Required,
+					stFromType,
+					"in",
+					"",
+					"",
+					indent+"\t\t",
+					c.updateFieldMap(fieldMap, toSubIdentifier),
+					append(prevKeyPrefixes, toSubIdentifier),
+				)
+				if err != nil {
+					return err
+				}
+
+				c.append(indent+"\t\t", "return")
+				c.append(indent+"\t", "}") // end of function definition
+
+				c.append(indent+"\t", "out.", keyPrefix+pascalCase(toField.Name), " = ", converterMethodName, "(", fromIdentifier, ")")
+
+				delete(c.convStructMap, toFieldType.Name)
```

Comment:
why the delete

Reply:
The delete is necessary since a converter method might have been generated somewhere deep, which wouldn't be accessible if the same type conversion was needed again somewhere less deep

Reply:
More importantly, the generated method is not safe for re-use because that conversion and the current one might not have the same code due to fieldMap overrides